app/controllers/app_controller.rb

```diff
@@ -0,0 +1,126 @@
+# frozen_string_literal: true
+
+class AppController < BaseController
+  def initialize(request, session)
+    super(request, session)
+
+    @player, @game = session.load if session.present?
+
+    @response_data = {}
+  end
+
+  def index
+    return redirect_to :game if @session.present?
+
+    show_page :index
+  end
+
+  def rules
+    return redirect_to :game if @session.present?
+
+    show_page :rules
+  end
+
+  def statistics
+    return redirect_to :game if @session.present?
+
+    @response_data[:statistics] = Database.load
+
+    show_page :statistics
+  end
+
+  def game
+    return show_page :game if @session.present?
+
+    return redirect_to :root unless params? && validate_player
+
+    initialize_player
+    initialize_game
+
+    @session.save(player: @player, game: @game)
+
+    show_page :game
+  end
+
+  def submit_answer
+    return redirect_to :root unless @session.present?
+
+    obtain_guess_result
+
+    return redirect_to :game if @game.errors.include?(CodebreakerDiz::GuessFormatError)
+
+    @session.save(player: @player, game: @game)
+
+    return redirect_to :game_results if game_finished?
+
+    show_page :game
+  end
+
+  def hint
+    return redirect_to :root unless @session.present?
```

Comment:
Let's add a custom rack middleware for checking if the user is authorized and remove these duplications:
`return redirect_to :root unless @session.present?`
The middleware should return 403 status